ansible_base/resource_registry/apps.py

```diff
@@ -103,6 +114,33 @@ def connect_resource_signals(sender, **kwargs):
             signals.post_save.connect(handlers.update_resource, sender=cls)
             signals.post_delete.connect(handlers.remove_resource, sender=cls)
 
+            if _should_reverse_sync():
+                # Wrap save() in a transaction and sync to resource server
+                cls._original_save = cls.save
+
+                # Avoid late binding issues
+                def save(self, *args, _original_save=cls._original_save, **kwargs):
+                    with transaction.atomic():
+                        # We need to know if this is a new object before we save it
+                        action = "create" if self._state.adding else "update"
+                        # Save so we get an ansible_id if it's a new object
+                        _original_save(self, *args, **kwargs)
+                        # Send the object to the resource server
+                        handlers.sync_to_resource_server(self, action)
```

Comment:
Why this, as opposed to just connecting your method to the `post_save` signal? Or even adding it on to the existing `handlers.update_resource` that already runs post-save?

You can pass a value from pre_save to post_save if you really need to... just stashing it on the object if you really need to.

That won't allow you to wrap it in a transaction, but I'd argue that's the responsibility of the app using DAB, and wrapping things in another transaction (really a save-point) is only going to sow more chaos.

Reply:
See this commit: https://github.com/ansible/django-ansible-base/pull/548/commits/e60196c850b02c03cf073ce8ca1c0962d9cdf0f8

Using a signal was the initial approach. As you mention, we can't wrap it in a transaction which is what we're really trying to do here. The original approach got close - it opened a transaction in pre_save, and closed it in post_save. But if post_save never gets called (if `save()` fails), then the transaction would never get closed.

So I moved to this approach of monkeypatching `save()` instead to wrap it in a transaction.

> but I'd argue that's the responsibility of the app using DAB

That puts a lot of onus on each app vs just doing it in one place and letting each app inherit the behavior.

> is only going to sow more chaos

Can you elaborate a bit here?

Reply:
There's no benefit to nesting more than 1 transaction in a request, because the outcome can only be that it rolls back and returns an error code. I had to introduce a transaction for RBAC checking of related permissions for a few reasons:

https://github.com/ansible/django-ansible-base/blob/95f88964e2cea5af0712c91d9de758aa6f4651e4/ansible_base/rbac/api/related.py#L123-L125

I probably shouldn't have done it that way, but I knew where it would be used, eda-server, which was not enabling the ATOMIC_TRANSACTIONS setting, and AWX is. I believe it would be better that everywhere _DAB_ wants to introduce a transaction (including the code I wrote myself that I linked there), it first checks to see if it is already in a transaction, and if it is, doesn't.

But also, assume your transaction here is necessary, and think about what that implies. That means you POST to create an object and get a 400. Sure, that _could_ mean that we left a resource inconsistency. But there's a _much more basic_ bug here - that a request returning a 400 still created the object. There's really not much wiggle room there. That's a bug.

Reply:
> it first checks to see if it is already in a transaction

Okay, we can check `connection.in_atomic_block`. It might not always be reliable (I think we can be in a transaction beyond being in an atomic block), but it might help in some or even most cases.

> that a request returning a 400 still created the object

I'm not sure I understand this, I think I need a concrete example. I'm assuming you mean if we get a 400 back from the resource server when we try to replicate the object to it. But that should always mean the request failed and the object didn't get saved on the resource server. Is there a case where that isn't true? I think I'm misunderstanding.

Reply:
> I'm not sure I understand this, I think I need a concrete example. I'm assuming you mean if we get a 400 back from the resource server when we try to replicate the object to it.

No, that's not what I'm talking about. Imagine there is no resource syncing at all.

Say you make a POST to create an organization in eda-server, for an example. Now imagine that you get an error code, that could be 400, 500, or something else. But not a 201. After getting that response, check the organization list. Was the organization created?

This question isn't irrelevant, because it relates strongly to the "why are you using a transaction?" question.

Reply:
Well yes that would be a bug in the app. But I would argue that _if_ the app is broken and creates the org anyway, then it should also still do so in the resource server. The whole point here is to maintain _internal_ consistency across the cluster. Even if we do something stupid, let's make sure that stupid thing is synced.

Reply:
I added a commit that looks at `connection.in_atomic_block` before deciding to wrap the block in a transaction. It's a util function that we can use elsewhere, as well.

Reply:
> But I would argue that if the app is broken and creates the org anyway, then it should also still do so in the resource server.

Then let's say it's a 500 error due to a `post_save` signal that errors. That, combined with requests being non-atomic, explain why the object gets created in the database when it shouldn't have.

But here's the thing - the object is incorrectly created, but it is also not synced to the resource server. The `post_save` signal fires as a part of `_original_save(self, *args, **kwargs)`, which happens before the syncing happens. Since it's an uncaught exception it doesn't ever bother syncing the object. It's created in the local server and not present in the resource server. That's _with_ the change here.

It's also possible that it had an error in the view or serializer finalization logic, in which case it would get synced to the resource server. There's no consistency in any case.

So I'm still lacking an original reason for the transaction here.

Reply:
I'm really confused and missing something.

If post_save fails, the [original transaction from save()](https://github.com/django/django/blob/main/django/db/models/base.py#L989) will get rolled back, and the exception will get re-raised. At that point the outer transaction - the one I'm adding here, will see the exception that was re-raised and rollback and (itself) reraise, and that will happen before we ever call `sync_to_resource_server()`, right?

Here's my test:

1. Replace all of `sync_to_resource_server()` with this:
    ```python
    def sync_to_resource_server(instance, action):
        raise exception("sync_to_resource_server threw this")
    ```
2. In test_app:
```
In [1]: def bad_post_save(sender, **kwargs):
   ...:     raise Exception("oh no")
   ...: 

In [2]: from django.db.models.signals import post_save

In [3]: post_save.connect(bad_post_save, Organization)

In [4]: o = Organization.objects.first()

In [5]: o.save()
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
Cell In[5], line 1
----> 1 o.save()

File ~/dev/ansible/aap-gateway/django-ansible-base/ansible_base/resource_registry/apps.py:128, in connect_resource_signals.<locals>.save(self, _original_save, *args, **kwargs)
    126 action = "create" if self._state.adding else "update"
    127 # Save so we get an ansible_id if it's a new object
--> 128 _original_save(self, *args, **kwargs)
    129 # Send the object to the resource server
    130 sync_to_resource_server(self, action)

File ~/dev/ansible/aap-gateway/django-ansible-base/ansible_base/lib/abstract_models/common.py:155, in AbstractCommonModel.save(self, *args, **kwargs)
    152     field_value = getattr(self, field, None)
    153     setattr(self, field, ansible_encryption.encrypt_string(field_value))
--> 155 return super().save(*args, **kwargs)

File ~/dev/ansible/aap-gateway/django-ansible-base/ansible_base/lib/abstract_models/common.py:133, in CreatableModel.save(self, *args, **kwargs)
    130     if self.created_by is None:
    131         self.created_by = current_user_or_system_user()
--> 133 return super().save(*args, **kwargs)

File ~/dev/ansible/aap-gateway/django-ansible-base/ansible_base/lib/abstract_models/common.py:101, in ModifiableModel.save(self, *args, **kwargs)
     98     if has_update_fields:
     99         kwargs['update_fields'] = update_fields
--> 101 return super().save(*args, **kwargs)

File ~/dev/ansible/dab-venv/lib/python3.12/site-packages/django/db/models/base.py:814, in Model.save(self, force_insert, force_update, using, update_fields)
    811     if loaded_fields:
    812         update_fields = frozenset(loaded_fields)
--> 814 self.save_base(
    815     using=using,
    816     force_insert=force_insert,
    817     force_update=force_update,
    818     update_fields=update_fields,
    819 )

File ~/dev/ansible/dab-venv/lib/python3.12/site-packages/django/db/models/base.py:892, in Model.save_base(self, raw, force_insert, force_update, using, update_fields)
    890 # Signal that the save is complete
    891 if not meta.auto_created:
--> 892     post_save.send(
    893         sender=origin,
    894         instance=self,
    895         created=(not updated),
    896         update_fields=update_fields,
    897         raw=raw,
    898         using=using,
    899     )

File ~/dev/ansible/dab-venv/lib/python3.12/site-packages/django/dispatch/dispatcher.py:177, in Signal.send(self, sender, **named)
    170 if (
    171     not self.receivers
    172     or self.sender_receivers_cache.get(sender) is NO_RECEIVERS
    173 ):
    174     return []
    176 return [
--> 177     (receiver, receiver(signal=self, sender=sender, **named))
    178     for receiver in self._live_receivers(sender)
    179 ]

Cell In[1], line 2, in bad_post_save(sender, **kwargs)
      1 def bad_post_save(sender, **kwargs):
----> 2     raise Exception("oh no")

Exception: oh no

2024-08-06 19:22:15,205 DEBUG    []  asyncio Using selector: EpollSelector
In [6]: 
 
```

The exception in `sync_to_resource_server()` is never thrown because the function is never called.

So I think I still don't quite understand the issue. I'm trying to, I promise.